File: flutter/android/src/main/kotlin/com/clevertap/flutter/nativedisplay/CleverTapNativeDisplayPlugin.kt
package com.clevertap.flutter.nativedisplay

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class CleverTapNativeDisplayPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel

    // Set by the host app after initialising CleverTap Core SDK.
    // The plugin delegates all display-unit calls to this bridge.
    var bridge: NativeDisplayPluginBridge? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "fetchDisplayUnit" -> {
                val unitId = call.argument<String>("unitId")
                if (unitId == null) {
                    result.error("INVALID_ARGUMENT", "unitId is required", null)
                    return
                }
                val json = bridge?.fetchDisplayUnit(unitId)
                if (json != null) result.success(json) else result.success(null)
            }
            "pushViewedEvent" -> {
                val unitId = call.argument<String>("unitId")
                if (unitId != null) bridge?.pushViewedEvent(unitId)
                result.success(null)
            }
            "pushClickedEvent" -> {
                val unitId = call.argument<String>("unitId")
                val elementId = call.argument<String?>("elementId")
                if (unitId != null) bridge?.pushClickedEvent(unitId, elementId)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    companion object {
        const val CHANNEL = "com.clevertap.flutter.nativedisplay"
    }
}

interface NativeDisplayPluginBridge {
    // Return display unit JSON string for the given unitId, or null if not found.
    fun fetchDisplayUnit(unitId: String): String?

    // Report viewed event to CleverTap Core SDK.
    fun pushViewedEvent(unitId: String)

    // Report clicked event to CleverTap Core SDK.
    fun pushClickedEvent(unitId: String, elementId: String?)
}

File: flutter/ios/Classes/CleverTapNativeDisplayPlugin.swift
import Flutter
import UIKit

public class CleverTapNativeDisplayPlugin: NSObject, FlutterPlugin {

    // Set by the host app after initialising CleverTap Core SDK.
    public var bridge: NativeDisplayPluginBridge?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.clevertap.flutter.nativedisplay",
            binaryMessenger: registrar.messenger()
        )
        let instance = CleverTapNativeDisplayPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Arguments missing", details: nil))
            return
        }

        switch call.method {
        case "fetchDisplayUnit":
            guard let unitId = args["unitId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "unitId is required", details: nil))
                return
            }
            result(bridge?.fetchDisplayUnit(unitId: unitId))

        case "pushViewedEvent":
            if let unitId = args["unitId"] as? String {
                bridge?.pushViewedEvent(unitId: unitId)
            }
            result(nil)

        case "pushClickedEvent":
            if let unitId = args["unitId"] as? String {
                let elementId = args["elementId"] as? String
                bridge?.pushClickedEvent(unitId: unitId, elementId: elementId)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

public protocol NativeDisplayPluginBridge {
    // Return display unit JSON string for the given unitId, or nil if not found.
    func fetchDisplayUnit(unitId: String) -> String?

    // Report viewed event to CleverTap Core SDK.
    func pushViewedEvent(unitId: String)

    // Report clicked event to CleverTap Core SDK.
    func pushClickedEvent(unitId: String, elementId: String?)
}
